File: ui/src/lib/clipboard.ts
/**
 * Reading a paste, without trusting the webview's copy of the clipboard.
 *
 * A macOS pasteboard can hold several items at once, and apps use that for a
 * multi-row selection: copying a block of build errors out of Xcode writes one
 * item per row. WebKit fills a paste event's `clipboardData` from the *first*
 * item only, so `getData("text/plain")` hands the page one row and drops the
 * rest — the whole clipboard is there, the webview just never asks for it. A
 * clipboard that also carries file URLs is worse: the event arrives typed as
 * "Files" with no text at all. Pasting into a native text field takes a
 * different path inside WebKit and gets everything, which is why a bare
 * terminal, or a round trip through another app (which collapses the
 * pasteboard to one item), pastes the full text. That is AGE-41.
 *
 * xterm.js pastes what `clipboardData` gives it, so a terminal pane asks the
 * backend for the real pasteboard instead (`read_clipboard_text` reads every
 * item) and falls back to the webview's text if that read comes up empty.
 */

import { readClipboardText } from "../api";

/**
 * The full clipboard text for a paste. `fromWebview` is the paste event's own
 * `text/plain`, used when the backend has nothing better — it is the truncated
 * copy, but on a single-item clipboard the two agree, and it is all there is
 * if the native read fails.
 */
export async function fullClipboardText(
  fromWebview: string,
  read: () => Promise<string | null> = readClipboardText,
): Promise<string> {
  try {
    const native = await read();
    return native || fromWebview;
  } catch {
    return fromWebview;
  }
}

/**
 * The clipboard as text for a paste the UI performs itself (a menu item, say)
 * rather than one the webview hands over. Same order: the real pasteboard,
 * then the webview's own reader — which needs a user gesture and can be
 * refused outright, so its rejection is left to propagate to the caller.
 */
export async function clipboardText(
  read: () => Promise<string | null> = readClipboardText,
): Promise<string> {
  const native = await read().catch(() => null);
  return native || (await navigator.clipboard.readText());
}

File: crates/agency-app/src/clipboard.rs
//! Reading the clipboard as *whole* text.
//!
//! macOS pasteboards can hold several items at once — copying a multi-row
//! selection out of an app like Xcode's issue navigator writes one item per
//! row. WebKit's paste path only ever reads item 0 (`Pasteboard::read` takes a
//! single index and defaults it to zero), so a paste inside the webview gets
//! the first row and silently drops the rest. Pasting the same clipboard into
//! a native app that reads every item — TextEdit, Terminal — yields all of it,
//! and re-copying from there collapses the pasteboard to one item, which is
//! why the round trip "fixes" it (AGE-41).
//!
//! So the frontend does not trust the webview's copy of the clipboard: it asks
//! here instead, and this reads every item and joins them with newlines, which
//! is the text the user selected.

/// Every plain-text item on the clipboard, newline-joined. `None` when the
/// clipboard holds no text at all (an image, say), which tells the caller to
/// fall back to whatever the webview handed it.
#[tauri::command]
pub fn read_clipboard_text() -> Option<String> {
    read_all()
}

#[cfg(target_os = "macos")]
fn read_all() -> Option<String> {
    use objc2_app_kit::{NSPasteboard, NSPasteboardTypeString};

    // NSPasteboard is documented as safe to use from any thread.
    let pasteboard = NSPasteboard::generalPasteboard();
    let items = pasteboard.pasteboardItems();
    let parts: Vec<String> = items
        .into_iter()
        .flatten()
        .filter_map(|item| unsafe { item.stringForType(NSPasteboardTypeString) })
        .map(|s| s.to_string())
        .collect();
    if !parts.is_empty() {
        return Some(parts.join("\n"));
    }
    // No per-item string: fall back to the whole-pasteboard read, which also
    // covers promised/lazy flavors an item enumeration can miss.
    unsafe { pasteboard.stringForType(NSPasteboardTypeString) }.map(|s| s.to_string())
}

#[cfg(not(target_os = "macos"))]
fn read_all() -> Option<String> {
    // The multi-item truncation is a macOS pasteboard/WebKit behavior; elsewhere
    // the webview's own clipboard read is authoritative.
    None
}
